validator/server_arb/redis/consumer.go

```diff
@@ -0,0 +1,131 @@
+package redis
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/offchainlabs/nitro/pubsub"
+	"github.com/offchainlabs/nitro/util/redisutil"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+	"github.com/offchainlabs/nitro/validator"
+	"github.com/offchainlabs/nitro/validator/server_api"
+	"github.com/offchainlabs/nitro/validator/valnode/redis"
+)
+
+type ExecutionSpawner struct {
+	stopwaiter.StopWaiter
+	spawner validator.ExecutionSpawner
+
+	// consumers stores moduleRoot to consumer mapping.
+	consumers map[common.Hash]*pubsub.Consumer[*server_api.GetLeavesWithStepSizeInput, []common.Hash]
+	config    *redis.ValidationServerConfig
+}
+
+func NewExecutionSpawner(cfg *redis.ValidationServerConfig, spawner validator.ExecutionSpawner) (*ExecutionSpawner, error) {
+	if cfg.RedisURL == "" {
+		return nil, fmt.Errorf("redis url cannot be empty")
+	}
+	redisClient, err := redisutil.RedisClientFromURL(cfg.RedisURL)
+	if err != nil {
+		return nil, err
+	}
+	consumers := make(map[common.Hash]*pubsub.Consumer[*server_api.GetLeavesWithStepSizeInput, []common.Hash])
+	for _, hash := range cfg.ModuleRoots {
+		mr := common.HexToHash(hash)
+		c, err := pubsub.NewConsumer[*server_api.GetLeavesWithStepSizeInput, []common.Hash](redisClient, server_api.RedisBoldStreamForRoot(cfg.StreamPrefix, mr), &cfg.ConsumerConfig)
+		if err != nil {
+			return nil, fmt.Errorf("creating consumer for validation: %w", err)
+		}
+		consumers[mr] = c
+	}
+	return &ExecutionSpawner{
+		consumers: consumers,
+		spawner:   spawner,
+		config:    cfg,
+	}, nil
+}
+
+func (s *ExecutionSpawner) Start(ctx_in context.Context) {
```

Comment:
I think this could be embedded into the "normal" redis consumer, possibly as a a separate StartBoldSpawner function that does the same and is called from the consumer's Start.